Adding parameterizations of Hurwitz matrices

Alex-Hache opened the pull request:
I added a parameterizations of Hurwitz matrices, i.e. the one that have eigenvalues with negative real parts they represent stable linear dynamical systems. Optionally one can enforce a specific bound on the matrix decay rate named alpha, it is related to the concept of alpha-stability in Lyapunov linear theory.

In order to build this parameterization, I changed the constructor of the Skew class that now takes a size parameter. 
Now the parameterization is instantiated with _register_manifold instead of register_parameterization.
The test_skey.py file has been modified accordingly.

(This is my first PR ever do not hesitate to come back to me if something does not meet the requirements)

Alex-Hache committed:
Adding Hurwitz parameterization with a lower bound on the decay rate of matrix eigenvalues, modifying skew class to implement a right inverse and using _register_manifold_ instead of register_parameterization. Now skew constructor needs a size argument.

Alex-Hache committed:
Fixing tolerances and removing batch support

Alex-Hache committed:
remove prints

lezcano commented:
Oh, the first external contribution to the project!

Sadly, the project is a bit out of date, but given that it's useful for people, I'll try to spend some time one of these days bringing it up to speed (supporting new PyTorch versions, perhaps even torch.compile, updating the action runners, etc)

This may take a bit of time, and if I don't answer in a couple weeks feel free to ping me again, but yeah, let's upgrade geotorch a bit!

Alex-Hache commented:
Alright I won't hesitate to reach out in a few weeks, you can contact me if you want me to do some adjustments on the package I'll be happy to help !

Alex-Hache commented:
Hello, Mario ! 

I wonder if you have any spare time in following weeks to work on the project ?

I may be very interested in making the library work with torch.compile but it seems the current main issue is the cache manager but it is so convenient for me which is training rnns that I do not know if it is worth to drop it for compile.

Do you want me to look on it ?

Cheers

Alexandre

lezcano commented on geotorch/constraints.py:
```diff
@@ -550,3 +550,28 @@ def positive_semidefinite_fixed_rank(
+    r"""Adds an alpha_stability parametrization to the matrix ``module.tensor_name``.
+
+    When accessing ``module.tensor_name``, the module will return the parametrized
+    version :math:`A` so that :math:`\min_{\lambda \in Sp(A)} \Re(\lambda)<= -\alpha`.
```
Note that we don't use the spectrum notation here. Can you rewrite this following the notation we use in other manifolds where we discuss eigenvalues=

lezcano replied:
nit. "Absolute value of the upper bound on the real part of the eigevenalues of :math: `X`", but otherwise LGTM

lezcano commented on geotorch/constraints.py:
```diff
@@ -550,3 +550,28 @@ def positive_semidefinite_fixed_rank(
+            on which the parametrization will be applied. Default: ``"weight"``
+        alpha (float): Bound on the decay rate and eigevenalues of matrix A
+    """
+    return _register_manifold(module, tensor_name, Hurwitz, alpha)
```
Add an end of line (the linter should have fixed this, but it's not well maintained...)

lezcano commented on geotorch/hurwitz.py:
```diff
@@ -0,0 +1,164 @@
+    def __init__(self, size, alpha: float = 0.0, triv="expm"):
+        r"""
+        Manifold of matrices with eigenvalues with negative real parts,
+        also called Hurwitz matrices. They represend linear stable linear dynamical systems.
```
```suggestion
        also called Hurwitz matrices.
```
no need to discuss applications in the docs.

lezcano commented on geotorch/hurwitz.py:
```diff
@@ -0,0 +1,164 @@
+        Manifold of matrices with eigenvalues with negative real parts,
+        also called Hurwitz matrices. They represend linear stable linear dynamical systems.
+
+        We have the following result : A is Hurwitz with prescribed decay rate \alpha
```
Decay rate has not been defined. Please write self-contained definitions.

lezcano commented on geotorch/hurwitz.py:
```diff
@@ -0,0 +1,164 @@
+
+        We have the following result : A is Hurwitz with prescribed decay rate \alpha
+        if and only if it can be written as
+        :math:`A = P^{-1}(-\frac{Q}{2} +S) - \alpha I_n`
```
```suggestion
        :math:`A = P^{-1}(-\frac{Q}{2} +S) - \alpha \mathrm{I}_n`
```

lezcano commented on geotorch/hurwitz.py:
```diff
@@ -0,0 +1,164 @@
+        We have the following result : A is Hurwitz with prescribed decay rate \alpha
+        if and only if it can be written as
+        :math:`A = P^{-1}(-\frac{Q}{2} +S) - \alpha I_n`
+        with $P > 0, Q > 0$ and ^S^skew-symmetric
```
What does `P > 0` mean here? Do you mean `P \in PSD(n)`? Please follow the notation in the rest of the library.

lezcano replied:
Also, Fix the skew-symmetric notation.

lezcano commented on geotorch/hurwitz.py:
```diff
@@ -0,0 +1,164 @@
+        with $P > 0, Q > 0$ and ^S^skew-symmetric
+        Args:
+            size (torch.size): Size of the tensor to be parametrized
+            alpha (float): the upper bound on the matrix's eigenvalues real part
```
This should be defined in the main docs, and perhaps here just say it's the upper bound in absolute value of the eigen values or smth

Alex-Hache replied:
Here is the new docstring : 

        r"""
        Manifold of matrices with eigenvalues with negative real parts,
        also called Hurwitz matrices.

        We have the following result : A is Hurwitz with prescribed decay rate \alpha
        if and only if it can be written as
        ..math::
            A = P^{-1}(-\frac{Q}{2} +S) - \mathrm{I}_n
        with :math:`P \in \operatorname{PSD}(n), Q \in \operatorname{PSD}(n)` and :math: `S \in \operatorname{Skew}(n)`

        Args:
            size (torch.size): Size of the tensor to be parametrized
            alpha (float): the upper bound on the matrix's eigenvalues real part
            triv (str or callable): Optional.
                A map that maps skew-symmetric matrices onto the orthogonal matrices
                surjectively. This is used to optimize the :math:`Q` in the eigenvalue
                decomposition. It can be one of ``["expm", "cayley"]`` or a custom
                callable. Default: ``"expm"``
        """
        
    I am not yet very familiar with how Sphinx works so I added \oepratorname to define the manifold as you did in the docs.

Alex-Hache replied:
When you mean in the main docs do you want me to write something in Sphinx somewhere in the docs too ?

lezcano replied:
by "the main doc" I meant "not in the Args part, but above it".

The docs above LGTM, just drop the "   We have the following result :" part, and surround in :math: the first A in "A is Hurwitz with prescribed..."

lezcano replied:
Side note, generally it's better to just push a commit with the changes, as it's easier for me to review them in github rather than in the comments.

lezcano commented on geotorch/hurwitz.py:
```diff
@@ -0,0 +1,164 @@
+        """
+        with torch.no_grad():
+
+            X_p = torch.empty((self.n, self.n))
```
Any chance you can reuse the samplers for each of the other manifolds and simply compose them via the trivialisation?

Alex-Hache replied:
Honestly very good question ! In terms of code I think we can yes indeed.
In terms of sampling I do not know... my knowledge of random matrix theory is limited and I have not found (yet !) a result on sampling in this manifold. I'll change the code to use the previous samplers.

lezcano replied:
There is not much random matrix theory involved here. The way we sample in geotorch is by sampling on the base spaces and mapping these samples by the trivialisation, so I'd hope this can be done compositionally (to be fair, now that I think about it, we should have been able to write this generically, in the base class. Not sure why I didn't do this).

Anyway, could you do this for this one class?

In terms of what this means mathematically, you are just choosing the pushforward of the measures on the base manifolds along the trivialisation. Not that this is super important here, but yeah.

lezcano commented on geotorch/skew.py:
```diff
@@ -32,10 +45,57 @@ def forward(self, X):
+                *(self.tensorial_size + (self.n, self.n)))
+            init_(X)
+            if lower:
+                tril_X = X.tril(-1)
```
might want to do this in-place iwth `tril_`? Same for the `triu_` below.

lezcano reviewed (commented):
First review, will do another pass this weekend

Alex-Hache committed:
Improved docstrings for constraints.alpha_stable
and Hurwitz class. Hurwitz.sample now uses PSD.sample() and Skew.sample() internally.

lezcano commented on geotorch/hurwitz.py:
```diff
@@ -18,16 +18,17 @@ class Hurwitz(ProductManifold):
+        also called Hurwitz matrices.
 
-        We have the following result : A is Hurwitz with prescribed decay rate \alpha
+        :math`A` is Hurwitz with prescribed decay rate \alpha
```
```suggestion
        :math`A` is Hurwitz with prescribed decay rate :math:`\alpha`
```

lezcano commented on geotorch/hurwitz.py:
```diff
@@ -131,26 +134,16 @@ def sample(self, init_=torch.nn.init.xavier_normal_, gamma=1.0):
-
-            
-            P_inv = torch.inverse(P)
             A = P_inv @ (-0.5 * Q + S) - self.alpha * self.In
```
can you simply call `self.submersion` here?

lezcano commented:
tensorial_size is not supported since the resolution of the vectorized Lyapunov equation is annoying the kroenecker product 

I am not sure how this is related? `tensorial_size` is just the batched version of the equation, so the extension is always trivial for any manifold (think `vmap` in jax or just samplign in `M^n`)

Alex-Hache committed:
Update geotorch/hurwitz.py

Co-authored-by: Mario Lezcano Casado <3291265+lezcano@users.noreply.github.com>

Alex-Hache committed:
Added batched version support, indeed vmap will do the trick for korenecker product.

lezcano commented on geotorch/hurwitz.py:
```diff
@@ -0,0 +1,172 @@
+                I_single = torch.eye(self.n, device=A.device, dtype=A.dtype)
+                return torch.kron(I_single, A_single) + torch.kron(A_single, I_single)
+            
+            flat_batch_size = torch.prod(torch.tensor(self.tensorial_size)) if self.tensorial_size else 1
```
```suggestion
            flat_batch_size = math.prod(self.tensorial_size)
```

lezcano commented on test/test_hurwitz.py:
```diff
@@ -0,0 +1,106 @@
+    def test_register_hurwitz(self):
+        layer = torch.nn.Linear(self.size[-2], self.size[-1])
+        geotorch.alpha_stable(layer, "weight", alpha=0.5)
+        self.assertTrue(torch.all(torch.real(torch.linalg.eigvals(layer.weight)) <= -0.5))
```
please add an empty line at the end of the file. In general, did you remember to run the linter?

Alex-Hache replied:
I did not run it before but I did now (I work with autopep8)

lezcano replied:
The linter in the project is `black` so please use that one.
The CI is broken, which is not great, but you can see here the command that should return zero:
https://github.com/lezcano/geotorch/blob/ba38d406c245d609fee4b4dac3f6427bf6d73a8e/.github/workflows/build.yml#L48C1-L48C31

lezcano reviewed (commented):
Looking good, just two minor points.
Tomorrow I'll do a final pass, and I think we can merge. Thank you!

Alex-Hache committed:
Runing Linter and some fixes for the batched version of submersion inv

Alex-Hache committed:
Running flake8 and black linters

lezcano commented on examples/eigenvalue.py:
```diff
@@ -3,6 +3,7 @@
 of a symmetric matrix via the Rayleigh quotient, restricting the optimisation
 problem to the Sphere
 """
+
```
I guess this was not intended?

lezcano reviewed (approved):
At last I had time to look into this and it looks great!
Thank you for the contribution! I just left a tiny perf comment, but otherwise LGTM

Alex-Hache committed:
update submersion for perf

Alex-Hache commented on geotorch/hurwitz.py:
```diff
@@ -0,0 +1,167 @@
+        return PSD(size, triv=triv), PSD(size, triv=triv), Skew(size)
+
+    def submersion(self, Q, P, S):
+        return P @ (-0.5 * Q + S) - self.alpha * self.In
```
Make sense, did the modifs. Thanks again for the help ! If I have time this summer I'll do another PR for fixed trace matrices.
